src/main/scala/util/Replacement.scala

```diff
@@ -37,43 +63,147 @@ class SeqRandom(n_ways: Int) extends SeqReplacementPolicy {
+
+    // Compute next value of triangular matrix
+    // set the referenced way as more recent than every other way
+    nextState.zipWithIndex.map { case (e, i) =>
+      e := Mux(i.U === way, 0.U(n_ways.W), moreRecentVec(i) | wayDec)
+    }
+
+    nextState.zipWithIndex.tail.foldLeft((nextState.head.apply(n_ways-1,1),0)) { case ((pe,pi),(ce,ci)) => (Cat(ce.apply(n_ways-1,ci+1), pe), ci) }._1
+  }
+
   def access(way: UInt) {
-    state_reg := get_next_state(state_reg,way)
+    state_reg := get_next_state(state_reg, way)
   }
-  def access(ways: Seq[ValidIO[UInt]]) {
-    state_reg := ways.foldLeft(state_reg)((prev, way) => Mux(way.valid, get_next_state(prev, way.bits), prev))
+  def access(ways: Seq[Valid[UInt]]) {
+    when (ways.map(_.valid).orR) {
+      state_reg := get_next_state(state_reg, ways)
+    }
+    for (i <- 1 until ways.size) {
+      cover(PopCount(ways.map(_.valid)) === i.U, s"LRU_UpdateCount$i", s"LRU Update $i simultaneous")
+    }
   }
-  def get_next_state(state: UInt, way: UInt) = {
-    var next_state = state << 1
-    var idx = UInt(1,1)
-    for (i <- log2Up(n)-1 to 0 by -1) {
-      val bit = way(i)
-      next_state = next_state.bitSet(idx, !bit)
-      idx = Cat(idx, bit)
+
+  def get_replace_way(state: UInt): UInt = {
+    val moreRecentVec = extractMRUVec(state)  // reconstruct lower triangular matrix
+    // For each way, determine if all other ways are more recent
+    val mruWayDec     = (0 until n_ways).map { i =>
+      val upperMoreRecent = (if (i == n_ways-1) true.B else moreRecentVec(i).apply(n_ways-1,i+1).andR)
+      val lowerMoreRecent = (if (i == 0)        true.B else moreRecentVec.map(e => !e(i)).reduce(_ && _))
+      upperMoreRecent && lowerMoreRecent
     }
-    next_state.extract(n-1, 1)
+    OHToUInt(mruWayDec)
   }
-  def replace = get_replace_way(state_reg)
-  def get_replace_way(state: UInt) = {
-    val shifted_state = state << 1
-    var idx = UInt(1,1)
-    for (i <- log2Up(n)-1 to 0 by -1) {
-      val in_bounds = Cat(idx, UInt(BigInt(1) << i))(log2Up(n)-1, 0) < UInt(n)
-      idx = Cat(idx, in_bounds && shifted_state(idx))
+
+  def way = get_replace_way(state_reg)
+  def miss = access(way)
+  def hit = {}
+}
+
+class PseudoLRU(n_ways: Int) extends ReplacementPolicy {
+  //example bits storage format for 4-way PLRU:
+  // [2] = ways 3-2 older than ways 1-0
+  // [1] = way 3 older than way 2
+  // [0] = way 1 older than way 0
+  def nBits = n_ways - 1
+  private val state_reg = Reg(UInt(nBits.W))
+  def state_read = WireDefault(state_reg)
+
+  def access(way: UInt) {
+    state_reg := get_next_state(state_reg, way)
+  }
+  def access(ways: Seq[Valid[UInt]]) {
+    when (ways.map(_.valid).orR) {
+      state_reg := get_next_state(state_reg, ways)
+    }
+    for (i <- 1 until ways.size) {
+      cover(PopCount(ways.map(_.valid)) === i.U, s"PLRU_UpdateCount$i", s"PLRU Update $i simultaneous")
+    }
+  }
+
+  def get_next_state(state: UInt, way: UInt, this_ways: Int): UInt = {
```

Comment:
I'm a fan of descriptive parameter and variable names. Would either n_ways or this_nways be better than just "this_ways", which (to me at least) does not clearly convey that it is the total number of ways?

Similarly, I'd change "way" to "touched_way", and if typing that is too annoying, say val w = touched_way, and use w in the code below for conciseness. That way, people fully understand the parameters to get_next_state() even without any comments.

Reply:
I changed it to `tree_nways` and `touch_way`